File: src/ipc/tmdb.ts
import { invoke } from "@tauri-apps/api/core";
import {
  TmdbEpisodeDetail,
  TmdbMovieDetail,
  TmdbMovieResult,
  TmdbSeasonSummary,
  TmdbTrendingItem,
  TmdbTvDetail,
  TmdbTvResult,
} from "../types";

export async function searchTmdbMovies(query: string, year?: number): Promise<TmdbMovieResult[]> {
  return await invoke<TmdbMovieResult[]>("search_tmdb_movies", { query, year });
}

export async function getMovieDetails(movieId: number): Promise<TmdbMovieDetail> {
  return await invoke<TmdbMovieDetail>("get_movie_details", { movieId });
}

export async function searchTv(query: string, year?: number): Promise<TmdbTvResult[]> {
  return await invoke<TmdbTvResult[]>("search_tv", { query, year });
}

export async function getTvDetails(seriesId: number): Promise<TmdbTvDetail> {
  return await invoke<TmdbTvDetail>("get_tv_details", { seriesId });
}

export async function getTvSeasons(seriesId: number): Promise<TmdbSeasonSummary[]> {
  return await invoke<TmdbSeasonSummary[]>("get_tv_seasons", { seriesId });
}

export async function getTvEpisodes(
  seriesId: number,
  seasonNumber: number
): Promise<TmdbEpisodeDetail[]> {
  return await invoke<TmdbEpisodeDetail[]>("get_tv_episodes", { seriesId, seasonNumber });
}

export async function getTrending(mediaType?: "movie" | "tv"): Promise<TmdbTrendingItem[]> {
  return await invoke<TmdbTrendingItem[]>("get_trending", { mediaType });
}

export async function getPopularMovies(): Promise<TmdbMovieResult[]> {
  return await invoke<TmdbMovieResult[]>("get_popular_movies");
}

export async function getPopularTv(): Promise<TmdbTvResult[]> {
  return await invoke<TmdbTvResult[]>("get_popular_tv");
}

File: src-tauri/src/commands/tmdb.rs
use std::sync::Arc;
use tauri::State;

use crate::error::AppError;
use crate::services::tmdb_service::{
    TmdbEpisodeDetail, TmdbMovieDetail, TmdbMovieResult, TmdbSeasonSummary, TmdbService,
    TmdbTrendingItem, TmdbTvDetail, TmdbTvResult,
};

#[tauri::command]
pub async fn search_tmdb_movies(
    query: String,
    year: Option<u16>,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbMovieResult>, AppError> {
    tmdb_service.search_movies(&query, year).await
}

#[tauri::command]
pub async fn get_movie_details(
    movie_id: i64,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<TmdbMovieDetail, AppError> {
    tmdb_service.get_movie_details(movie_id).await
}

#[tauri::command]
pub async fn search_tv(
    query: String,
    year: Option<u16>,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbTvResult>, AppError> {
    tmdb_service.search_tv(&query, year).await
}

#[tauri::command]
pub async fn get_tv_details(
    series_id: i64,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<TmdbTvDetail, AppError> {
    tmdb_service.get_tv_details(series_id).await
}

#[tauri::command]
pub async fn get_tv_seasons(
    series_id: i64,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbSeasonSummary>, AppError> {
    let details = tmdb_service.get_tv_details(series_id).await?;
    Ok(details.seasons)
}

#[tauri::command]
pub async fn get_tv_episodes(
    series_id: i64,
    season_number: u32,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbEpisodeDetail>, AppError> {
    tmdb_service.get_tv_episodes(series_id, season_number).await
}

#[tauri::command]
pub async fn get_trending(
    media_type: Option<String>,
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbTrendingItem>, AppError> {
    let m = media_type.unwrap_or_else(|| "movie".to_string());
    tmdb_service.get_trending(&m).await
}

#[tauri::command]
pub async fn get_popular_movies(
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbMovieResult>, AppError> {
    tmdb_service.search_movies("popular", None).await
}

#[tauri::command]
pub async fn get_popular_tv(
    tmdb_service: State<'_, Arc<TmdbService>>,
) -> Result<Vec<TmdbTvResult>, AppError> {
    tmdb_service.search_tv("popular", None).await
}
